documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,284 @@
+        false, // RequestRundown
+        256 // CircularBufferMB
+    );
+    EventPipeEventSource source = new EventPipeEventSource(session.EventStream);
+
+    source.Dynamic.All += (TraceEvent obj) => {
+        Console.WriteLine(obj.EventName);
+    }
+    try
+    {
+        source.Process();
+    }
+    catch (EventStreamException) {}
+    // NOTE: This exception does not currently exist. It is something that needs to be added to TraceEvent.
+}
+```
+
+#### 2. Trigger a core dump when CPU usage goes above a certain threshold
+```cs
+
+using Microsoft.Diagnostics.NETCore.Client;
+
+public void TriggerDumpOnCpuUsage(int processId, int threshold)
+{
+    List<EventPipeProvider> providers = new List<EventPipeProvider>() {
+        new EventPipeProvider(
+            "System.Runtime",
+            EventLevel.Informational,
+            (long)ClrTraceEventParser.Keywords.None,
+            new Dictionary<string, string>() {
+                { "EventCounterIntervalSec", "1" }
+            }
+        ),
+        new EventPipeProvider(
+            "Microsoft-DotNETCore-SampleProfiler",
+            EventLevel.Informational
+        ),
+    };
+    DiagnosticsClient client = new DiagnosticsClient(processId);
+    EventPipeSession session = client.StartEventPipeSession(providers);
+    EventPipeEventSource source = new EventPipeEventSource(session.EventStream)
+    source.Dynamic.All += (TraceEvent obj) => {
+        if (obj.EventName.Equals("EventCounters"))
+        {
+            // I know this part is ugly. But this is all TraceEvent.
+            IDictionary<string, object> payloadFields = (IDictionary<string, object>)(obj.GetPayloadValueByName("Payload"));
+            if (payloadFields["Name"].ToString().Equals("cpu-usage"))
+            {
+                double cpuUsage = Double.Parse(payloadFields["Mean"]);
+                if (cpuUsage > (double)threshold)
+                {
+                    DiagnosticsClient.WriteDump(processId, "/tmp/minidump.dmp", DumpType.Normal);
+                }
+            }
+        }
+    }
+    try
+    {
+        source.Process();
+    }
+    catch (EventStreamException) {}
+}
+```
+
+#### 3. Trigger a CPU trace for given number of seconds
+```cs
+
+using Microsoft.Diagnostics.NETCore.Client;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Task;
+
+public void TraceProcessForDuration(int processId, int duration, string traceName)
+{
+    List<EventPipeProvider> cpuProviders = new List<EventPipeProvider>() {
+        new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Informational, ClrTraceEventParser.Keywords.Default),
+        new EventPipeProvider("Microsoft-DotNETCore-SampleProfiler", EventLevel.Informational, ClrTraceEventParser.Keywords.None)
+    };
+
+    Task copyTask = Task.Run(async () =>
+    {
+        using(FileStream fs = new FileStream(traceName, FileMode.Create, FileAccess.Write))
+        {
+            await traceSession.stream.CopyToAsync(fs);
```

Comment:
Is stream really lowercase?

Reply:
No, I think we renamed it "EventStream" and the sample needs to be updated.